### PR TITLE
ensure we only delete old tokens on force logout

### PR DESCRIPTION
**Pull request by paynejacob**
https://github.com/rancher/rancher/issues/34149

If the leader pod crashes before finishing the force logout or saving the configuration map users could be logged out twice or not at all.  This change ensures that all users will always be logged out on upgrade and any user that logs in after the upgrade will not be logged out again.

### CODE REVIEW

**Comment by paynejacob** [comment]
I should also note I was unable to reproduce the senario that QA outlines in the ticket.  This is the only logical way I could see users being logged out twice.

File: pkg/rancher/migrations.go [hunk 1]
```diff
@@ -39,15 +45,33 @@ func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, to
...
 			return nil
 		}
 	}
 
 	logrus.Infof("Detected %s upgrade, forcing logout for all web users", migrationVersion)
 
+	// save that we are starting a new migration
+	lastMigrationStart = time.Now().UnixNano()
+	cm.Data[rancherVersionKey] = rancherversion.Version
+	cm.Data[migrationStartedKey] = fmt.Sprintf("%d", lastMigrationStart)
```
Comment (snasovich) [inline pkg/rancher/migrations.go:64]: Why not `strconv.Itoa`? Same below for completion time.

**Review by snasovich — DISMISSED** [review]
One minor comment, otherwise LGTM

File: pkg/rancher/migrations.go [hunk 2]
```diff
@@ -39,15 +45,33 @@ func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, to
...
-		if lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=") {
+		if (lastMigration == "dev" || version.Compare(migrationVersion, lastMigration, "<=")) && lastMigrationStart < lastMigrationCompleted {
 			return nil
 		}
 	}
 
 	logrus.Infof("Detected %s upgrade, forcing logout for all web users", migrationVersion)
 
+	// save that we are starting a new migration
+	lastMigrationStart = time.Now().UnixNano()
```
Comment (dramich) [inline pkg/rancher/migrations.go:62]: You get smaller numbers with just `Unix`
```suggestion
	lastMigrationStart = time.Now().Unix()
```

File: pkg/rancher/migrations.go [hunk 3]
```diff
@@ -39,15 +45,33 @@ func forceUpgradeLogout(configMapController controllerv1.ConfigMapController, to
 		}
 	}
 
-	// we only want to migrate if the previous installation was earlier than the migration version
+	// parse out the start and end time
+	lastMigrationStart, _ := strconv.ParseInt(cm.Data[migrationStartedKey], 10, 64)
```
Comment (dramich) [inline pkg/rancher/migrations.go:49]: Use the helpers for all the conversions https://pkg.go.dev/strconv#hdr-Numeric_Conversions
`Atoi`
`Itoa`
It keeps it consistent and it's much faster than using sprintf which relies on reflection.

File: pkg/rancher/migrations.go (hunk 2 above)
Comment (paynejacob) [reply pkg/rancher/migrations.go:62]: I am concerned this will finish in less than a second.

Comment (dramich) [reply pkg/rancher/migrations.go:62]: I highly doubt that, especially if you are calling out to update the config. However, just update your time check to include `=`

File: chart/templates/secret.yaml [hunk 4]
```diff
@@ -6,8 +6,8 @@ metadata:
 type: Opaque
 data:
   {{- $existingSecret := lookup "v1" "Secret" .Release.Namespace "bootstrap-secret" }}
-  {{- if $existingSecret }}
-  bootstrapPassword: {{ $existingSecret.data.bootstrapPassword }}
-  {{- else }}
-  bootstrapPassword: {{ .Values.bootstrapPassword | b64enc | quote }}
-  {{- end }}
\ No newline at end of file
+    {{- if and $existingSecret $existingSecret.data }}
```
Comment (StrongMonkey) [inline chart/templates/secret.yaml:9]: Would recommend to put this into a separate PR. If anything go wrong reverting might be easier.

Comment (dramich) [reply chart/templates/secret.yaml:9]: I agree, separate pr as it's not part of this fix or issue.

Comment (paynejacob) [reply chart/templates/secret.yaml]: https://github.com/rancher/rancher/pull/34249

**Comment by paynejacob** [comment]
Closing this pr as the solution has changed dramatically and this conversation is no longer applicable.
